src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs

```diff
@@ -0,0 +1,391 @@
+            private int[] _lengths;
+
+            public AndroidTzData()
+            {
+                string tzFilePath = GetTimeZoneDirectory() + TimeZoneFileName;
+                using (FileStream fs = File.OpenRead(tzFilePath))
+                {
+                    ReadHeader(tzFilePath, fs);
+                }
+            }
+
+            [MemberNotNull(nameof(_ids))]
+            [MemberNotNull(nameof(_byteOffsets))]
+            [MemberNotNull(nameof(_lengths))]
+            private unsafe void ReadHeader(string tzFilePath, Stream fs)
+            {
+                int size = Math.Max(Marshal.SizeOf(typeof(AndroidTzDataHeader)), Marshal.SizeOf(typeof(AndroidTzDataEntry)));
+                Span<byte> buffer = stackalloc byte[size];
+                AndroidTzDataHeader header = ReadAt<AndroidTzDataHeader>(tzFilePath, fs, 0, buffer);
+
+                header.indexOffset = NetworkToHostOrder(header.indexOffset);
+                header.dataOffset = NetworkToHostOrder(header.dataOffset);
+
+                // tzdata files are expected to start with the form of "tzdata2012f\0" depending on the year of the tzdata used which is 2012 in this example
+                if (header.signature[0] != (byte)'t' || header.signature[1] != (byte)'z' || header.signature[2] != (byte)'d' || header.signature[3] != (byte)'a' || header.signature[4] != (byte)'t' || header.signature[5] != (byte)'a' || header.signature[11] != 0)
+                {
+                    var b = new StringBuilder();
+                    for (int i = 0; i < 12; ++i) {
+                        b.Append(' ').Append((header.signature[i]).ToString("x2"));
+                    }
+
+                    throw new InvalidOperationException(SR.Format(SR.InvalidOperation_BadTZHeader, tzFilePath, b.ToString()));
+                }
+
+                ReadIndex(tzFilePath, fs, header.indexOffset, header.dataOffset, buffer);
+            }
+
+            [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2087:UnrecognizedReflectionPattern",
+                Justification = "Implementation detail of Android TimeZone")]
+            private unsafe T ReadAt<T>(string tzFilePath, Stream fs, long position, Span<byte> buffer)
+                where T : struct
+            {
+                int size = Marshal.SizeOf(typeof(T));
+                Debug.Assert(buffer.Length >= size);
+
+                fs.Position = position;
+                int numBytesRead;
+                if ((numBytesRead = fs.Read(buffer)) < size)
+                {
+                    throw new InvalidOperationException(SR.Format(SR.InvalidOperation_ReadTZError, tzFilePath, position, size, numBytesRead, size));
+                }
+
+                fixed (byte* b = buffer)
+                {
+                    return (T)Marshal.PtrToStructure((IntPtr)b, typeof(T))!; // Is ! the right way to handle Unboxing a possibly null value. Should there be some check instead?
+                }
```

Comment:
This should not use `Marshal.PtrToStructure` at all, and instead read directly into the `T`. It will also avoid the need for the caller to provide temporary buffer of the right size. The linker annotation won't be needed either.

Something like this:

```
private static unsafe T ReadAt<T>(string tzFilePath, Stream fs, long position) where T : unmanaged
{
    T value;
    Span<byte> buffer = new Span<byte>(&value, sizeof(T));
    
    fs.Position = position;

    int numBytesRead;
    while ((numBytesRead = fs.Read(buffer)) > 0)
    {
        buffer = buffer.Slice(numBytesRead);
        if (buffer.IsEmpty)
            break;
    }

    if (!buffer.IsEmpty)
    {
        throw new InvalidOperationException(...);
    } 

    return value;
}
```

Reply:
The part I don't like about this approach is that the bytes coming from the file are not directly "blittable" into .NET objects because of endian issues. The file is always big endian. So now you have to remember that 

```C#
            private unsafe struct AndroidTzDataHeader
            {
                public fixed byte signature[12]; // "tzdata2012f\0"
                public int indexOffset;
                public int dataOffset;
                public int finalOffset;
            }
```

After you read the bytes directly into that structure, those offset fields are not the correct `int`s. But instead they are reversed endian when running on little endian machines.

I think a cleaner approach is to read the portion of the file into a `Span<byte>` buffer, and then read the bytes out imperatively, like we do for TZifHead:

https://github.com/dotnet/runtime/blob/4c92aef2b08f9c4374c520e7e664a44f1ad8ce56/src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.cs#L1758-L1784

This also has the advantage that you don't need to store the `signature` in `AndroidTzDataHeader` at all. You can just interpret the buffer directly. You honestly probably don't even need the structs at all with this approach, since we are never storing them for anything.

Reply:
I agree that reading the bytes out imperatively looks better in situations like this one.